### PR TITLE
splunk configuration changes 

### PR DESCRIPTION
**Pull request by ts3ng**
Changes after testing vpc peering with secops configurations to allow connections from coreos to security splunk servers.

### CODE REVIEW

**Comment by ts3ng** [comment]
these changes are dependent of PR https://github.com/adobe-platform/ethos-systemd/pull/12.  @matthewdfuller @chr0n1x

File: v1/opt/splunk/README.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#Security Splunk FleetUnit
```
Comment (chr0n1x) [inline v1/opt/splunk/README.md]: nice!

File: v1/opt/splunk/setup/common/splunk.sh [hunk 2]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/bash -x
+source /etc/environment
+DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+source $DIR/lib/helpers.sh
+SPLUNK_FORWARD_SERVER_LIST=$(etcd-get /splunk/SPLUNK_FORWARD_SERVER_LIST)
```
Comment (chr0n1x) [inline v1/opt/splunk/setup/common/splunk.sh]: these configs should follow the `/splunk/config` format.

e.g.: `/splunk/config/forward-server-list`

File: v1/opt/splunk/setup/common/splunk.sh [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+SPLUNK_DIR="/opt/splunk/etc/system/local"
+
+#create splunk configuration directory
+mkdir -p $SPLUNK_DIR
+#generate/fetch secure certs in configurations directory
+sudo echo -n "$(etcd-get /splunk/SPLUNK_ADOBECAAS_CERT | awk '{gsub(/\\n/,"\n")}1')" > $SPLUNK_DIR/test_adobeacaas_cert
+sudo echo -n "$(etcd-get /splunk/SPLUNK_CA | awk '{gsub(/\\n/,"\n")}1')" > $SPLUNK_DIR/test_splunk_ca
+
+#generate configurtion outputs file
+sudo echo "[tcpout]
```
Comment (chr0n1x) [inline v1/opt/splunk/setup/common/splunk.sh]: you should try using bash heredocs here.

File: v1/opt/splunk/splunk-journald.service [hunk 4]
```diff
@@ -4,26 +4,26 @@ Requires=docker.service
...
 EnvironmentFile=/etc/environment
+Environment="INSTANCEHOSTNAME=curl -s http://169.254.169.254/latest/meta-data/hostname"
 User=core
 Restart=always
 TimeoutStartSec=0
 ExecStartPre=/usr/bin/systemctl is-active bootstrap
-ExecStartPre=/usr/bin/bash -c \
-"if [[ -f /etc/profile.d/etcdctl.sh ]]; then source /etc/profile.d/etcdctl.sh;fi && \
-docker pull $($IMAGE)"
+ExecStartPre=/usr/bin/docker pull index.docker.io/adobeplatform/docker-splunk:latest
```
Comment (chr0n1x) [inline v1/opt/splunk/splunk-journald.service]: why not have this in `etcd get /images`?

File: v1/opt/splunk/splunk-journald.service [hunk 5]
```diff
@@ -4,26 +4,26 @@ Requires=docker.service
...
 ExecStartPre=-/usr/bin/docker kill splunk-journald
 ExecStartPre=-/usr/bin/docker rm -f splunk-journald
 ExecStart=/usr/bin/bash -c \
 "if [[ -f /etc/profile.d/etcdctl.sh ]]; then source /etc/profile.d/etcdctl.sh; fi && \
 sudo -E /usr/bin/docker run --name splunk-journald \
---hostname=${COREOS_PRIVATE_IPV4} \
+-v /opt/splunk/etc/system/local:/var/opt/splunk/etc/system/local \
```
Comment (chr0n1x) [inline v1/opt/splunk/splunk-journald.service]: any reason why the host dir is so nested?

File: v1/opt/splunk/splunk-journald.service (hunk 4 above)
Comment (matthewdfuller) [inline v1/opt/splunk/splunk-journald.service]: It should be - and it should be set as a default in `v1/setup/leader/001-defaults.sh`

File: v1/opt/splunk/splunk-journald.service [hunk 6]
```diff
@@ -4,26 +4,26 @@ Requires=docker.service
...
+--hostname=$($INSTANCEHOSTNAME) \
+--log-opt max-size=`etcdctl get /docker/config/logs-max-size` \
+--log-opt max-file=`etcdctl get /docker/config/logs-max-file` \
```
Comment (matthewdfuller) [inline v1/opt/splunk/splunk-journald.service]: I don't think this is required any longer since Chris made it a default. @f4tq can you confirm?

File: v1/opt/splunk/splunk-journald.service (hunk 4 above)
Comment (matthewdfuller) [inline v1/opt/splunk/splunk-journald.service]: Also, this should be defined at the top as:

```
Environment="IMAGE=`etcdctl get /images/splunk`
```

Then, it can be used here as: `$($IMAGE)`

File: v1/opt/splunk/splunk-journald.service [hunk 7]
```diff
@@ -4,26 +4,26 @@ Requires=docker.service
...
 -p 1514:1514/udp \
 -e SPLUNK_START_ARGS="--accept-license" \
--e SPLUNK_FORWARD_SERVER=`etcdctl get /splunk/SPLUNK_FORWARD_SERVER` \
--e SPLUNK_ADD='udp 1514 -sourcetype journald -hostname ${COREOS_PRIVATE_IPV4} -index os -connection_host none' \
-$($IMAGE)"
+-e SPLUNK_ADD='udp 1514 -sourcetype journald -hostname '$($INSTANCEHOSTNAME)' -index os -connection_host none' \
+adobeplatform/docker-splunk:latest"
```
Comment (matthewdfuller) [inline v1/opt/splunk/splunk-journald.service]: This should be `$($IMAGE)`

File: v1/opt/splunk/splunk-journald.service (hunk 4 above)
Comment (ts3ng) [inline v1/opt/splunk/splunk-journald.service]: @chr0n1x @matthewdfuller 
I had images originally but that's when we had setup scripts awhile back ago before everything was shuffled around.  It's not diff than https://github.com/adobe-platform/mesos-systemd/blob/master/v3/opt/sumologic/sumologic-journald.service.   I don't have no preference where it's set if we want to put all image ref in etcd i'm fine with that, should i put it in the setup/common/splunk.sh?  it should be set only when the optional unit is enabled right?

File: v1/opt/splunk/splunk-journald.service (hunk 5 above)
Comment (ts3ng) [inline v1/opt/splunk/splunk-journald.service]: That's how splunk fowarder looks for the configurations.  installation is in /opt/splunk.  I want to leave it 1:1 mapping on the subdir on the coreos host so there is some consistency on how splunk is configured (ops people can ref splunk docs for info).  There are several folders in the splunk configuration settings if you want more detail on splunk configurations see https://docs.splunk.com/Splexicon:Configurationfile.

Comment (chr0n1x) [inline v1/opt/splunk/splunk-journald.service]: ah ok cool 👍

File: v1/opt/splunk/splunk-journald.service (hunk 6 above)
Comment (f4tq) [inline v1/opt/splunk/splunk-journald.service]: True.  You don't need the etcdctl log-opt max-{size,file} call.  The infrastructure PR took care of that.

I need to change the PR submission documentation.

**Comment by ts3ng** [comment]
okay all changes suggested pushed, thanks for the feedback guys! 
@f4tq @chr0n1x @matthewdfuller

**Comment by chr0n1x** [comment]
👍

**Comment by matthewdfuller** [comment]
👍 
However, I can't merge until I merge in my other PR. I will update here once I do.

**Comment by matthewdfuller** [comment]
@ts3ng copying my comment from Slack for visibility:

PR has a merge conflict, but one other thing too: `v1/opt/splunk/splunk-journald.service` should be in `v1/opt/splunk/fleet/splunk-journald.service`

**Comment by ts3ng** [comment]
@matthewdfuller sorry didnt' notice the conflict went ahead and merged those changes over looks better now.

**Comment by matthewdfuller** [comment]
@ts3ng Thanks - just move that last file into the `fleet` folder and we'll be good

**Comment by ts3ng** [comment]
@matthewdfuller I think the merge already moved them to /opt/splunk/fleet/ since there were no conflicts on the file.

**Comment by matthewdfuller** [comment]
Ah, you're right, sorry! Will review one last time and then merge in a bit

File: v1/opt/splunk/fleet/splunk-journald.service [hunk 8]
```diff
@@ -4,25 +4,24 @@ Requires=docker.service
...
+Environment="IMAGE=etcdctl get /images/splunk"
+Environment="INSTANCEHOSTNAME=curl -s http://169.254.169.254/latest/meta-data/hostname"
 User=core
 Restart=always
 TimeoutStartSec=0
 ExecStartPre=/usr/bin/systemctl is-active bootstrap
-ExecStartPre=/usr/bin/bash -c \
-"if [[ -f /etc/profile.d/etcdctl.sh ]]; then source /etc/profile.d/etcdctl.sh;fi && \
-docker pull $($IMAGE)"
+ExecStartPre=/usr/bin/docker pull index.docker.io/adobeplatform/docker-splunk:latest
```
Comment (matthewdfuller) [inline v1/opt/splunk/fleet/splunk-journald.service]: Why did you remove this? The "image" should be defined in the Environment above, not hard coded

File: v1/opt/splunk/setup/common/splunk.sh [hunk 9]
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/bash -x
+source /etc/environment
+DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+source $DIR/lib/helpers.sh
```
Comment (matthewdfuller) [inline v1/opt/splunk/setup/common/splunk.sh]: This is not going to work - $DIR is now set to `v1/opt/splunk/setup/common`. Then, it sources that path + `helpers.sh`. But `helpers.sh` exists several levels up

File: v1/opt/splunk/setup/common/splunk.sh [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+SPLUNK_FORWARD_SERVER_LIST=$(etcd-get /splunk/config/forward_server_list)
```
Comment (matthewdfuller) [inline v1/opt/splunk/setup/common/splunk.sh]: For consistency with our other names, can you convert these to dashes? `forward-server-list`, etc

File: v1/opt/splunk/setup/common/splunk.sh [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+#create splunk configuration directory
+mkdir -p $SPLUNK_DIR
+#generate/fetch secure certs in configurations directory
+cat << EOF > /$SPLUNK_DIR/adobecaas.crt
+$(etcd-get /splunk/config/adobecaas_cert | awk '{gsub(/\\n/,"\n")}1')
+EOF
+
+cat << EOF > /$SPLUNK_DIR./genericForwarder.pem
+$(etcd-get /splunk/config/ca_cert | awk '{gsub(/\\n/,"\n")}1')
```
Comment (matthewdfuller) [inline v1/opt/splunk/setup/common/splunk.sh]: Same here `ca-cert`

**Comment by matthewdfuller** [comment]
A few more changes to make that I didn't catch originally. Additionally, the config values `/splunk/config/forward-server-list`, `ca-cert`, etc. need to be added to the `infrastructure` `secrets-example.json` file.

File: v1/opt/splunk/setup/common/splunk.sh (hunk 10 above)
Comment (matthewdfuller) [inline v1/opt/splunk/setup/common/splunk.sh]: Also, this is currently in infrastructure example config as `"/splunk/SPLUNK_FORWARD_SERVER": "ds.adobesecaas.com:8089",`

**Comment by ts3ng** [comment]
Suggested changes done, infra changes here... https://github.com/adobe-platform/infrastructure/pull/118. 
cc: @matthewdfuller

File: v1/opt/splunk/setup/common/splunk.sh (hunk 9 above)
Comment (ts3ng) [inline v1/opt/splunk/setup/common/splunk.sh]: thanks for this one, didn't catch that change

File: v1/opt/splunk/fleet/splunk-journald.service (hunk 8 above)
Comment (ts3ng) [inline v1/opt/splunk/fleet/splunk-journald.service]: doh, not sure how that one got changed actually.

File: v1/opt/splunk/setup/common/splunk.sh (hunk 10 above)
Comment (ts3ng) [inline v1/opt/splunk/setup/common/splunk.sh]: +1 for consistency, added changes.

**Comment by ts3ng** [comment]
@matthewdfuller okay i think that's everything...  i talked to ops guys yesterday they want to push this change after they push ed's klam-ssh stuff through, so the timeline would be planned on splunk SE sometime next week, so it'll be good to get this PR approved this week/early next week.

File: v1/opt/splunk/fleet/splunk-journald.service [hunk 12]
```diff
@@ -4,25 +4,24 @@ Requires=docker.service
...
+Environment="IMAGE=etcdctl get /images/splunk"
+Environment="INSTANCEHOSTNAME=curl -s http://169.254.169.254/latest/meta-data/hostname"
 User=core
 Restart=always
 TimeoutStartSec=0
 ExecStartPre=/usr/bin/systemctl is-active bootstrap
-ExecStartPre=/usr/bin/bash -c \
-"if [[ -f /etc/profile.d/etcdctl.sh ]]; then source /etc/profile.d/etcdctl.sh;fi && \
-docker pull $($IMAGE)"
+ExecStartPre=/usr/bin/docker pull $($IMAGE)
```
Comment (matthewdfuller) [inline v1/opt/splunk/fleet/splunk-journald.service]: This needs the whole `source /etc/profile.d/etcdctl.sh` from over there <<<. If this runs on the control, it will work fine, but any `etcd get` commands that run on worker or proxy will need to source etcdctl.sh first. The complete line should be:

```
ExecStartPre=/usr/bin/bash -c \
"if [[ -f /etc/profile.d/etcdctl.sh ]]; then source /etc/profile.d/etcdctl.sh;fi && \
docker pull $($IMAGE)"
```

**Comment by ts3ng** [comment]
haha sorry for all the changes @matthewdfuller, pls review one more time when you guys can spare the time, thanks!
cc: @chr0n1x if @matthewdfuller can't get to it this week.

**Comment by matthewdfuller** [comment]
👍  applies after a successful A/B with proxy/workers. @chr0n1x - can you review this and merge next week while I'm on PTO? Wait until @adobejmong and team merge the KLAM-SSH changes first.
